File: leetcode-js/158-read-n-characters-given-read4-ii-call-multiple-times.js
/**
 * Definition for read4()
 *
 * @param  {character[]} buf Destination buffer
 * @return {number}      The number of characters read
 * read4 = function(buf) {
 *     ...
 * };
 */

/**
 * @param  {function} read4()
 * @return {function}
 */
var solution = function(read4) {

  var temp = [];

  /**
   * @param  {character[]} buf Destination buffer
   * @param  {number}      n   Maximum number of characters to read
   * @return {number}      The number of characters read
   */
  return function read(buf, n) {
    var buf4;

    if (temp.length < n) {
      do {
        buf4 = [];
        read4(buf4);
        temp.push.apply(temp, buf4);
        if (buf4.length < 4) {
          break;
        }
      } while (temp.length < n);
    }

    var len = Math.min(temp.length, n);
    var cur = temp.splice(0, len);

    buf.push.apply(buf, cur);

    return len;
  };
};

var r = [];

console.log(solution((buf4) => 0)(r, 1));

File: leetcode-java/158-read-n-characters-given-read4-ii-call-multiple-times.java
/* The read4 API is defined in the parent class Reader4.
      int read4(char[] buf); */

public class Solution extends Reader4 {

    private int index = 0;
    private int count = 0;
    private char[] buffer = new char[4];

    /**
     * @param buf Destination buffer
     * @param n   Maximum number of characters to read
     * @return    The number of characters read
     */
    public int read(char[] buf, int n) {
        int i = 0;

        while (i < n) {
            if (index == 0) {
                count = read4(buffer);
            }

            if (count == 0) {
                break;
            }

            while (i < n && index < count) {
                buf[i] = buffer[index];
                i++;
                index++;
            }

            if (index == count) {
                index = 0;
            }
        }

        return i;
    }
}
